s3dol/base.py

```diff
@@ -0,0 +1,227 @@
+Supports relative paths to access files under a prefix ending with slash.
+Hints added to help setup credentials.
+
+bucket = S3Dol()[profile][bucket_name]
+bucket['level1/level2/test-key'] == bucket['level1/']['level2/']['test-key']
+"""
+from dataclasses import dataclass
+import functools
+import os
+from typing import Union
+import warnings
+
+import boto3
+import dol
+
+
+class S3DolException(Exception):
+    pass
+
+
+noCredentialsFound = S3DolException(
+    'No AWS credentials found. Configure your AWS credentials as environment variables or in ~/.aws/credentials. '
+    'See https://github.com/i2mint/s3dol/#set-up-credentials for more information.'
+)
+
+
+def get_aws_credentials():
+    aws_access_key = os.environ.get('AWS_ACCESS_KEY_ID')
+    aws_secret_key = os.environ.get('AWS_SECRET_ACCESS_KEY')
+    aws_region = os.environ.get('AWS_DEFAULT_REGION')
+    aws_session_token = os.environ.get('AWS_SESSION_TOKEN')
+
+    return {
+        'aws_access_key_id': aws_access_key,
+        'aws_secret_access_key': aws_secret_key,
+        'region_name': aws_region,
+        'aws_session_token': aws_session_token,
+    }
+
+
+def list_profile_names():
+    """
+    Return a list of available AWS profiles
+    """
+    session = boto3.session.Session()
+    available_profiles = session.available_profiles
+    # Check environment variables for default profile
+    if os.environ.get('AWS_ACCESS_KEY_ID') and os.environ.get('AWS_SECRET_ACCESS_KEY'):
+        available_profiles.append('environment variables')
+
+    return available_profiles
+
+
+def get_client(profile_name=None, endpoint_url=None, **session_kwargs):
+    """
+    Return a boto3 client for the specified profile
+    """
+    if profile_name is None:
+        return _find_default_credentials(endpoint_url=endpoint_url, **session_kwargs)
+
+    if profile_name == 'environment variables':
+        aws_credentials = get_aws_credentials()
+        if (
+            not aws_credentials['aws_access_key_id']
+            or not aws_credentials['aws_secret_access_key']
+        ):
+            raise S3DolException(
+                'Missing AWS credentials in environment variables: AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY'
+            )
+        _skw = {**session_kwargs, **aws_credentials}
+        session = boto3.Session(**_skw)
+    else:
+        # Get credentials from profile
+        session = boto3.Session(profile_name=profile_name, **session_kwargs)
+    client = session.client('s3', endpoint_url=endpoint_url)
+    return client
+
+
+def _find_default_credentials(endpoint_url=None, **session_kwargs):
+    try:
+        return get_client('environment variables', **session_kwargs)
+    except S3DolException:
+        pass
+    session = boto3.Session(**session_kwargs)
+    if session.get_credentials() is None:
+        raise noCredentialsFound
+    client = session.client('s3', endpoint_url=endpoint_url)
+    return client
+
+
+@dataclass
+class BaseS3BucketReader(dol.base.KvReader):
```

Comment:
Do we really need those 2 Base classes

Reply:
I wanted to keep key path deliminating separated from the s3 bucket access.  Up to you if you think it should be renamed or merged.  I think the final S3Dol class should have everything included and the other classes can be used if want less.

Reply:
Seems to make sense. 
Again, decide with @valentin-feron today.